components/suggestion/search_suggestion_tagged_input.jsx

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import SearchSuggestionInput from './search_suggestion_input';
+const exampleTags = ['from:', 'in:', 'on:', 'before:', 'after:'];
+
+export default class SearchSuggestionTaggedInput extends SearchSuggestionInput {
+    static propTypes = {
+        onClear: PropTypes.func,
+        onInput: PropTypes.func,
+        onCompositionStart: PropTypes.func,
+        onCompositionUpdate: PropTypes.func,
+        onCompositionEnd: PropTypes.func,
+        pretext: PropTypes.string,
+        onKeyDown: PropTypes.func,
+        value: PropTypes.string,
+        onFocus: PropTypes.func,
+    }
+    constructor(props) {
+        super(props);
+        this.state = {
+            value: props.value, pairs: [], remainder: '',
+        };
+    }
+
+    onKeyDown = (e) => {
+        const value = this.getInput().value;
+
+        if (e.key === 'Backspace' && !value) {
+            const pairs = [...this.state.pairs];
+            pairs.pop();
+            this.setState({
+                pairs,
+            }, () => {
+                this.handleChange();
+            });
+        }
+
+        if (this.props.onKeyDown) {
+            this.props.onKeyDown(e);
+        }
+    }
+
+    isTag = (str) => {
+        const lcStr = str.toLowerCase();
+        return exampleTags.some((tag) => lcStr.startsWith(tag));
+    }
+
+    includesTags = (str) => {
+        const lcStr = str.toLowerCase();
+        return exampleTags.some((tag) => lcStr.includes(tag));
+    }
+
+    determinePairs(value = this.props.value) {
+        const includesTags = this.includesTags(value);
+        let remainder;
+        let pairs = [];
+
+        if (includesTags) {
+            const split = value.split(/(?<!:) /);
```

Comment:
Also, it doesn't look like this handles things like quoted search phrases. We might need to build a more sophisticated parser to handle this since we might be getting outside the bounds of what can be done with a simple split.

[This code](https://github.com/mattermost/mattermost-webapp/blob/master/utils/text_formatting.tsx#L647) may be helpful, although it kind of does the opposite thing we want here since it specifically removes the flags from the search query.